src/electionguard/precompute_buffers.cpp

```diff
@@ -183,122 +193,265 @@ namespace electionguard
+        if (!triple_queue.empty()) {
+            result = std::move(triple_queue.front());
+            triple_queue.pop();
         }
 
         return result;
     }
 
-    std::unique_ptr<Triple> PrecomputeBufferContext::getTriple()
+    std::unique_ptr<TwoTriplesAndAQuadruple> PrecomputeBuffer::getTwoTriplesAndAQuadruple()
     {
-        unique_ptr<Triple> result = nullptr;
+        if (!twoTriplesAndAQuadruple_queue.empty()) {
+            return popTwoTriplesAndAQuadruple().value();
+        }
+
+        return createTwoTriplesAndAQuadruple(*publicKey);
+    }
 
-        // take a lock while we get the triples and a quadruple
-        std::lock_guard<std::mutex> lock(queue_lock);
+    std::optional<std::unique_ptr<TwoTriplesAndAQuadruple>>
+    PrecomputeBuffer::popTwoTriplesAndAQuadruple()
+    {
+        unique_ptr<TwoTriplesAndAQuadruple> result = nullptr;
+        std::lock_guard<std::mutex> lock(quad_queue_lock);
 
         // make sure there are enough in the queues
-        if (!getInstance().triple_queue.empty()) {
-            result = std::move(getInstance().triple_queue.front());
-            getInstance().triple_queue.pop();
+        if (!twoTriplesAndAQuadruple_queue.empty()) {
+            result = std::move(twoTriplesAndAQuadruple_queue.front());
+            twoTriplesAndAQuadruple_queue.pop();
         }
 
         return result;
     }
 
-    void PrecomputeBufferContext::empty_queues()
+    std::tuple<std::unique_ptr<Triple>, std::unique_ptr<Triple>>
+    PrecomputeBuffer::createTwoTriples(const ElementModP &publicKey)
     {
-        std::lock_guard<std::mutex> lock(queue_lock);
-        uint32_t triple_size = getInstance().triple_queue.size();
-        for (int i = 0; i < (int)triple_size; i++) {
-            getInstance().triple_queue.pop();
+        auto triple1 = make_unique<Triple>(publicKey);
+        auto triple2 = make_unique<Triple>(publicKey);
+        return std::make_tuple(move(triple1), move(triple2));
+    }
+    unique_ptr<TwoTriplesAndAQuadruple>
+    PrecomputeBuffer::createTwoTriplesAndAQuadruple(const ElementModP &publicKey)
+    {
+        auto triple1 = make_unique<Triple>(publicKey);
+        auto triple2 = make_unique<Triple>(publicKey);
+        auto quad = make_unique<Quadruple>(publicKey);
+        return make_unique<TwoTriplesAndAQuadruple>(move(triple1), move(triple2), move(quad));
+    }
+
+#pragma endregion
+
+#pragma region PrecomputeBufferContext
+
+    void PrecomputeBufferContext::clear()
+    {
+        if (getInstance()._instance != nullptr) {
+            getInstance()._instance->clear();
+            getInstance()._instance = nullptr;
         }
-        uint32_t twoTriplesAndAQuadruple_size = getInstance().twoTriplesAndAQuadruple_queue.size();
-        for (int i = 0; i < (int)twoTriplesAndAQuadruple_size; i++) {
-            getInstance().twoTriplesAndAQuadruple_queue.pop();
+    }
+
+    void PrecomputeBufferContext::initialize(const ElementModP &publicKey,
```

Comment:
now the semantics are such that when you call `initialize` we assume you really mean that you want to blow out the cache and start over, but when you call `start` with an elgamal key, we'll only start over if it's different. This implies there can be one "global" precompute context in memory at a time, but you are free to make as many PrecomputeBuffer instances as you want. Later, we'll refactor the encrypt api's and the encryption mediator/device to accept an instance rather than the static reference so that an encryption device or mediator could handle multiple elections at the same time. I'll make an issue ofr that.